### DISCUSSION TITLE
ESP32C3 network stack issues receiving ACK packets (and ARP requests) when low RAM available during TCP stream writes

### DISCUSSION THREAD

**Discussion by petewatts** [micropython/micropython#16209] [ESP32]
Hi MicroPython fans,

Firstly, I'm completely new to MicroPython (but not python), and semi-new to programming on small RAM-limited boards like the ESP32, so it's entirely possible there's something obvious I'm doing wrong :) I've searched the old forum plus this discussion area and not turned up anything similar, closest I've seen are some network issues where the ESP32 role is the other way round with it acting as the TCP server rather client, but then then those don't appear to address this issue. Apologies if this has been covered previously and I didn't recognise it.

I've recently embarked on a project to move some scripts running on a Raspberry PI at home to an ESP32. They connect to a couple of bluetooth devices, and send values received to an MQTT server.  I figured the ESP32 might be perfect for this, and seemed like the perfect opportunity to give MicroPython a go instead of writing in C/C++. It's highly likely I've a lot more to learn around how best to tame memory management carefully on MicroPython for these small boards.

Long story short, what I'm finding is, if I write too many messages to the MQTT server (basically packets to the network TCP connection) when the ESP32 has low RAM available, typically around 4KB reported by `gc.mem_free()`in my reproducible test, then the network stack doesn't successfully receive the ACK packet in response to the data being sent, and blocks the sync/asyncio task writing to the stream. In my larger application the issue seems to kick in with around 10KB reported by `gc.mem_free()`, the difference probably related to fragmentation in the heap. Whilst the ACK is ignored by ESP32, the network stack of the ESP32 continues the usual re-send of the last packet it hasn't received an acknowledgement to yet, but never sees the ACK that's then resent, so continues to be blocked. 

I'm guessing it's something to do with the network stack needing to allocate some block of memory which it can't in order to receive the response. When it's in this state, it also no longer responds to ARP requests, so the server side eventually looses the MAC address of the ESP32, so doesn't know where to respond with the ACK. To test this further, with a rather crude fudge in my test scenario, freeing up some memory does seem to allow the network stack to jump back into life, albeit with a lost packet in my test, but that packet loss could be the result of the exception I needed to cause to unblock the stream write so I could then free up some RAM, so the TCP connection probably became out of-sync, but the connection did then start processing more packets to issue a connection reset, so seems to show that it's probably related to a lack of heap space.

Bottom line currently, I can't see any easy way to spot when this situation occurs, which makes it difficult to take some action. If fact, I've observed in my larger application that many more bytes of data are successfully written to the stream in python, but never actually leave the ESP32 network stack, presumably due to it still waiting to receive the ACK from an earlier packet, and possibly related to heap allocation/fragmentation somehow allowing more writes to be buffered. I've also not managed to see any error messages from the ESP32 about the lack of heap space within the network stack when it's failing to receive the ACK. Is there some suggestions on how to debug this further?

This behaviour feels like it's related to the natural network back-off process to ensure downstream resources aren't overwhelm. However, I would have hoped the network stack would arrange to ensure writing a packet out to the network could successfully receive a response, even It if then blocks further writes. Seems like it causes itself a deadlock. Is there something I'm missing to easily detect or avoid up-front this situation occurring? How can I ensure there's enough RAM available for the network stack to successfully receive responses to the packets it's sending? Like I said above, it's highly likely I'm just missing some standard approach around reliably ensuring this situation can never occur. Any recommendations?

In my scenario, as everything is flowing towards the MQTT server, I would need to somehow detect this is occurring and then backoff if there's too many messages to send. I could restrict the number of readings in flight at any point in time, the tricky part is knowing how much free space (presumably relating to the largest allocatable heap space) I'm targeting to allow the network stack to work reliably. Currently I've been restricting this on the input side when I see memory errors recording the new values from the sensors, but that's too late by that point. Is there some typical or recommended  techniques used for this sort of thing that I might be missing?

In case it's useful, below is my current way of reproducing the issue. It's an overly simplified example purely to replicate the issue. Tested on the latest 1.24.0 released version for ESP32C3:

- `esptool.py --chip esp32c3 --port /dev/tty.usbmodem14201 erase_flash`
- `curl -L https://micropython.org/resources/firmware/ESP32_GENERIC_C3-20241025-v1.24.0.bin > mp-esp32c3-1.24.0.bin`
- `esptool.py --chip esp32c3 --port /dev/tty.usbmodem14201 --baud 460800 write_flash -z 0x0 mp-esp32c3-1.24.0.bin`
- `curl -L https://raw.githubusercontent.com/micropython/micropython-lib/refs/heads/master/micropython/umqtt.simple/umqtt/simple.py > umqtt-simple.py`
- `mpremote fs mkdir :lib`
- `mpremote fs mkdir :lib/umqtt`
- `mpremote fs cp umqtt-simple.py :lib/umqtt/simple.py`

Create the following script in `test_net.py`:
```
import gc
import network
from umqtt.simple import MQTTClient

wifi_ssid = "<ssid>"
wifi_password="<wifi-password>"
mqtt_server = "192.168.1.4"
mqtt_user = "<mqtt-user>"
mqtt_password = "<mqtt-password>"

def main():
    network.hostname("testnode")
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    if not wlan.isconnected():
        wlan.connect(wifi_ssid, wifi_password)
        while not wlan.isconnected():
            pass
    print('network config:', wlan.ifconfig())

    mqtt = MQTTClient("test-client", mqtt_server, user=mqtt_user, password=mqtt_password)
    mqtt.connect()

    gc.collect()
    print(f"Free: {gc.mem_free()}")
    allocations = []
    allocate = True
    while allocate:
        try:
            allocations.append(b"0" * 1024)
        except Exception as e:
            allocate = False

    count = 0
    while True:
        try:
            count += 1
            print(f"Sending {count} free={gc.mem_free()}")
            mqtt.publish(topic=b"testnet/ble-device/1234567890123456/my_sensor_reading", msg=f"Test message {count}")
        except Exception as e:
            print(f"MQTT run error: {e}")

main()
```

Run: `mpremote run test_net.py` gives the following output:
```
% mpremote run test_net.py
network config: ('192.168.1.120', '255.255.255.0', '192.168.1.254', '192.168.1.7')
Free: 154816
Sending 1 free=3968
```
where it then blocks/hangs after sending some number of messages to the mqtt topic, just 1 in the above example.

Monitoring the network on the mqtt server interface shows:
```
# tcpdump -i eth0 -n host 192.168.1.120
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
17:25:05.796526 IP 192.168.1.120.58360 > 192.168.1.4.1883: Flags [F.], seq 3994293256, ack 2131800658, win 5756, length 0
17:25:05.796883 IP 192.168.1.4.1883 > 192.168.1.120.58360: Flags [F.], seq 1, ack 1, win 32059, length 0
17:25:05.849457 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [S], seq 3640976949, win 5760, options [mss 1440], length 0
17:25:05.849735 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [S.], seq 379533917, ack 3640976950, win 32120, options [mss 1460], length 0
17:25:06.147632 IP 192.168.1.4.1883 > 192.168.1.120.58360: Flags [F.], seq 1, ack 1, win 32059, length 0
17:25:06.152681 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [.], ack 1, win 5760, length 0
17:25:06.154569 IP 192.168.1.120.58360 > 192.168.1.4.1883: Flags [.], ack 2, win 5755, length 0
17:25:06.156381 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 1:4, ack 1, win 5760, length 3
17:25:06.156519 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 4, win 32117, length 0
17:25:06.159324 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 4:13, ack 1, win 5760, length 9
17:25:06.159414 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 13, win 32108, length 0
17:25:06.161717 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 13:26, ack 1, win 5760, length 13
17:25:06.161803 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 26, win 32095, length 0
17:25:06.166555 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 26:59, ack 1, win 5760, length 33
17:25:06.166658 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 59, win 32062, length 0
17:25:06.167837 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [P.], seq 1:5, ack 59, win 32062, length 4
17:25:06.262581 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [.], ack 5, win 5756, length 0
17:25:06.350042 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 59:61, ack 5, win 5756, length 2
17:25:06.391608 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 61, win 32060, length 0
17:25:07.768085 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 59:61, ack 5, win 5756, length 2
17:25:07.768297 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 61, win 32060, length 0
17:25:10.764176 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 59:61, ack 5, win 5756, length 2
17:25:10.764325 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 61, win 32060, length 0
17:25:16.762392 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 59:61, ack 5, win 5756, length 2
17:25:16.762620 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 61, win 32060, length 0
17:25:28.763589 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 59:61, ack 5, win 5756, length 2
17:25:28.763806 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 61, win 32060, length 0
17:25:32.372178 ARP, Request who-has 192.168.1.120 tell 192.168.1.120, length 46
17:25:52.762126 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 59:61, ack 5, win 5756, length 2
17:25:52.762303 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 61, win 32060, length 0
17:25:57.987493 ARP, Request who-has 192.168.1.120 tell 192.168.1.4, length 28
17:25:59.011496 ARP, Request who-has 192.168.1.120 tell 192.168.1.4, length 28
17:26:00.035492 ARP, Request who-has 192.168.1.120 tell 192.168.1.4, length 28
17:26:32.371299 ARP, Request who-has 192.168.1.120 tell 192.168.1.120, length 46
17:26:40.761376 IP 192.168.1.120.58361 > 192.168.1.4.1883: Flags [P.], seq 59:61, ack 5, win 5756, length 2
17:26:40.761583 IP 192.168.1.4.1883 > 192.168.1.120.58361: Flags [.], ack 61, win 32060, length 0
17:26:45.859473 ARP, Request who-has 192.168.1.120 tell 192.168.1.4, length 28
17:26:46.883468 ARP, Request who-has 192.168.1.120 tell 192.168.1.4, length 28
17:26:47.907472 ARP, Request who-has 192.168.1.120 tell 192.168.1.4, length 28
17:27:32.370541 ARP, Request who-has 192.168.1.120 tell 192.168.1.120, length 46
```

where you can see `ack 61` being acknowledged but it looks like that never makes it back through the ESP32 network stack. You can also see the ARP requests being sent to the ESP32 and the ESP not sending responses.

Any help or direction on what I should do/try next would be very helpful. Is this intended behaviour? Am I doing something stupid? Is this a bug? Is there some tweak I should make to the MicroPython build to enable something?

Many thanks in advance for any help, direction, or pointers anyone can give.
Pete

**Comment by petewatts** [comment]
To follow up on my own question, I've found a lot of further information in the MicroPython docs that I hadn't fully seen on my first pass. A combination of `micropython.qstr_info()` and `gc.collect()` helps alleviate the issue.

Is there any way to convince the ESP32 network stack to pre-allocate the 1536 bytes it seems to need periodically? It would be great to have it allocate a little bit more RAM so that it has the resources it needs so that the deadlock situation can be avoided.

**Reply by projectgus** [reply]
> A combination of `micropython.qstr_info()` and `gc.collect()` helps alleviate the issue.

You mean `micropython.mem_info()`, yes?

The memory situation in ESP32 is a bit complex as there's a large ESP-IDF heap that belongs to the ESP32's RTOS operating system, and then the Python heap is a subset of this. When the Python heap runs out of space and can't free any even with a GC, it will grow by allocating more memory from the ESP-IDF heap.

The Wi-Fi and network stack both allocate from the ESP-IDF heap as well. However, they will start to drop packets (or Wi-Fi frames) if the ESP-IDF heap runs out.

Therefore, in your test posted above where you try to allocate new 1024-bytes object until allocations fails then at this point almost all of the available ESP-IDF heap will be used by Python and almost none will be available for ESP-IDF.

> Is there any way to convince the ESP32 network stack to pre-allocate the 1536 bytes it seems to need periodically? It would be great to have it allocate a little bit more RAM so that it has the resources it needs so that the deadlock situation can be avoided.

Unfortunately not. We were considering having a "tunable" amount of memory that MicroPython always leaves for ESP-IDF, but it's hard to actually make this work in practice (as you don't know how much RAM ESP-IDF has already used i.e. is it already at its maximum, versus how much memory it might need in the future.)

Keeping Python memory usage small should be enough to keep things running smoothly, though. If you're passing big chunks of buffers around then see if you can refactor the code to allocate the large buffer once at startup (maybe as a bytearray) and reuse it from then on. This is often enough to solve these kind of memory issues.

The other function that can help debug ESP32 memory issues is `esp32.idf_heap_info()`: https://docs.micropython.org/en/latest/library/esp32.html#esp32.idf_heap_info - although the output can take a moment to get your head around.
